### ISSUE TITLE
Correlation of input weights for different king positions/nets

### ISSUE THREAD

**Issue by ddobbelaere** [glinscott/nnue-pytorch#45]
I have been investigating the correlation between nets today and I would like to share my results (I don't really know a better place to put them than here :)).

First, denote the (coalesced) relevant input weights of a net as $w_{k, s, f}$, which is a real number for each king index k (from 0 to 63), piece square index s (from 0 to 623), and (hidden) feature index f (from 0 to 255).

In the sequel, we consider master net nn-62ef826d1a6d.nnue (an offspring from sergioviero's run) and nn-0f63c1539914.nnue (latest vdv net on fishtest).

Let's first investigate the correlation coefficient between the weights *of the same net* as of function of king index, defined as:

![image](https://user-images.githubusercontent.com/38748663/104817043-14bb9400-581f-11eb-9977-41f21e33d02a.png)

where we adopt the summation convention that <.> averages over all non-primed indices, so that the correlation coefficient corresponds to the usual definition and, if defined, leads to a positive semi-definite matrix (namely the covariance matrix of the input weights, normalized per king position).

Visually this leads to:
![master_king](https://user-images.githubusercontent.com/38748663/104817426-885ea080-5821-11eb-9585-7d477e785795.png)
![vdv_king](https://user-images.githubusercontent.com/38748663/104817429-8a286400-5821-11eb-824f-d559823b9125.png)

Observations:

- The "piece" weights are pretty correlated for different king indices (note the scale!).
- Both master and vdv net have lower correlation for indices 1 (square b1) and 6 (square g1). **I cannot explain this, any ideas?** EDIT: g1 corresponds to 0-0, b1 to 0-0-0 followed by prophylactic Kb1 :). Hmm, maybe it's just "black castles short" due to weird rotational symmetry atm.
- There are some other differences (other patterns) between master and vdv net that I cannot explain. Maybe related to coalescing (factorizer on)?

---

In the same way, let's define the correlation coefficient between the weights of two nets i and j (can be the same), as a function of the hidden feature index, as:

![image](https://user-images.githubusercontent.com/38748663/104817114-8d225500-581f-11eb-8c27-2797c25bca45.png)

where again <.> averages over all non-primed subindices.

Then we have the following correlation coefficient matrix between master and vdv features:

![cross](https://user-images.githubusercontent.com/38748663/104817440-9a404380-5821-11eb-9f5b-87449ba1e026.png)

Now, let's reorder the features so that they come in pairs, so as to maximize the correlation. Then we get this (note that the diagonal in the off-diagonal block matrices (corresponding to cross-terms $N^{(i,j)}_{f'} between nets, i.e., for different i and j) becomes visible and the all four block matrices look visually the same):

![cross_reordered](https://user-images.githubusercontent.com/38748663/104817444-9f9d8e00-5821-11eb-8eba-472b57b2c158.png)

Now, let's plot the input weights of the *reordered* nets (such that the most correlated features appear from left to right, top to bottom):

![master](https://user-images.githubusercontent.com/38748663/104816044-dd49e900-5818-11eb-9373-028cdbfa7e92.png)
![vdv](https://user-images.githubusercontent.com/38748663/104816045-df13ac80-5818-11eb-834d-5b435cea7bf8.png)

**Note that I have normalized the weights w.r.t. their max. value for each feature!**

Visually, some of the same "hidden feature recognizers" are distinguishable (same patterns), which is pretty amazing, considering the nets originate from totally different independent training runs (nodchip vs pytorch, but also other training data and parameters I guess).

If someone is interested, the code can be found here: https://github.com/ddobbelaere/nnue-pytorch/tree/correlation-research

**Comment by Sopel97** [comment]
I'm gonna refer only to the first 2 graphs. The first thing that comes to my mind is that different factorizers could cause some discrepancy there. The pytorch trainer doesn't factorize to HalfRelativeKP because it was worse in the initial tests, so there's no virtual features for "white queen 2 rank to right and 3 files up from the king" and stuff like that. This *might* explain the difference. The first (master) net shows correlation between nearby squares (a distance of 8 on the graph is a distance of 1 on the board). Such correlations makes sense, but are not really present in the net trained with the pytorch trainer. As a sidenote I have a hunch that one could forcefully get a net with feature correlations like the master net by using a factorizer for the king position with a gaussian kernel (so that feature `(our_king[f1, r1], piece[f2, r2]) is factorized into (our_king[f1, r1], piece[f2, r2]) + (our_king[f1+1, r1], piece[f2, r2])*a1 + (our_king[f1-1, r1], piece[f2, r2])*a2 + ...` where all factors are virtual features apart from the non-modified one).

As to why b1 and g1 seem to behave similarily is because with rotate180 transformation for black's perspective this corresponds to the king square after short castling. I'm not sure however why it's not correlated with weights from other king positions.

**Comment by ddobbelaere** [comment]
This is probably more some kind of research topic and not an issue per se :)
Also, it's not clear it still pertains to the HalfKAv2 architecture and current way of training.
